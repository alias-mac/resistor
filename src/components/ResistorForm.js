/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import _includes from 'lodash/includes';
import _find from 'lodash/find';

import React from 'react';
import Resistor from '../Resistor';

import '../styles/resistor-form.scss';

function BandSelector(props) {

  const { options, label, value, ...other } = props;

  // color is the mapping value or the first one of the list of options
  const color = value ? _find(options, { value }).className : options[0].className;

  return (
    <div className="form-group">
      <label>{label}
        <select {...other} value={value} className={`form-control ${color}`}>
          {options.map(option => (
            <option key={option.value} {...option}>{option.label}</option>
          ))}
        </select>
      </label>
    </div>
  );
}

BandSelector.propTypes = {
  options: React.PropTypes.arrayOf(React.PropTypes.shape({
    label: React.PropTypes.string,
    value: React.PropTypes.number,
  })).isRequired,
  label: React.PropTypes.string.isRequired,
  value: React.PropTypes.number.isRequired,
};

function dropDownMapper(def) {
  return {
    value: def.value,
    className: def.color,
    label: def.color,
  };
}

class ResistorForm extends React.Component {

  constructor(props) {
    super(props);

    this.onChange = this.onChange.bind(this);
  }

  onChange(event) {
    const selector = event.target;
    this.props.onChange(selector.name, selector.value);
  }

  render() {
    const { model, onChange, ...other } = this.props;

    const digitOptions = Resistor.significantDigits.map(dropDownMapper);
    const multiplierOptions = Resistor.multipliers.map(dropDownMapper);
    const toleranceOptions = Resistor.tolerances.map(dropDownMapper);
    const tempCoefsOptions = Resistor.tempCoefs.map(dropDownMapper);

    // normally the first digit isn't 0 (aka black) only if we have 0 ohms...
    const [, ...firstDigitOptions] = digitOptions;

    const bands = [];

    // only 5 and 6 has 3rd digit
    if (_includes([5, 6], model.get('bands'))) {
      bands.push(
        <BandSelector
          key="digit3" name="digit3" label="3rd Digit"
          value={model.get('digit3')}
          options={digitOptions}
          onChange={this.onChange}
        />,
      );
    }

    // all have at least 2 first digits
    bands.unshift(
      <BandSelector
        key="digit1" name="digit1" label="1st Digit"
        value={model.get('digit1')}
        options={firstDigitOptions}
        onChange={this.onChange}
      />,
      <BandSelector
        key="digit2" name="digit2" label="2nd Digit"
        value={model.get('digit2')}
        options={digitOptions}
        onChange={this.onChange}
      />,
    );

    // all have multiplier
    bands.push(
      <BandSelector
        key="multiplier" name="multiplier" label="Multiplier"
        value={model.get('multiplier')}
        options={multiplierOptions}
        onChange={this.onChange}
      />,
    );

    if (model.get('bands') !== 3) {
      bands.push(
        <BandSelector
          key="tolerance" name="tolerance" label="Tolerance"
          value={model.get('tolerance')}
          options={toleranceOptions}
          onChange={this.onChange}
        />,
      );
    }
    if (model.get('bands') === 6) {
      bands.push(
        <BandSelector
          key="tempCoef" name="tempCoef" label="Temp. Coef."
          value={model.get('tempCoef')}
          options={tempCoefsOptions}
          onChange={this.onChange}
        />,
      );
    }

    return (
      <form className="resistor-form" {...other}>
        <div className="row">
          <div className="form-group col-sm-6 offset-sm-3 col-md-2 offset-md-5">
            <label>Strips
              <select
                name="bands" className="form-control"
                value={model.get('bands')}
                onChange={this.onChange}
              >
                {[3, 4, 5, 6].map(i => (
                  <option key={i} value={i}>{i}-strip</option>
                ))}
              </select>
            </label>
          </div>
        </div>

        <div className="row">
          {bands.map((b, i) => {
            const offset = i ? '' : ` offset-sm-${6 - model.get('bands')}`;
            return (<div key={b.key} className={`col-xs-6 col-sm-2${offset}`}>{b}</div>);
          })}
        </div>
      </form>
    );
  }
}

ResistorForm.propTypes = {
  model: React.PropTypes.instanceOf(Resistor).isRequired,
  onChange: React.PropTypes.func,
};

ResistorForm.defaultProps = {
  onChange: () => { },
};

export default ResistorForm;
