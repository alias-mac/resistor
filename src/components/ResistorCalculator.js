/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import React from 'react';
import PropTypes from 'prop-types';

import '../styles/resistor-calc.scss';

import Resistor from '../Resistor';
import ResistorForm from './ResistorForm';
import ResistorValue from './ResistorValue';
import ResistorSvg from './ResistorSvg';

class ResistorCalculator extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      resistor: new Resistor({
        bands: 4,
        digit1: 1,
        digit2: 0,
        multiplier: 0,
      }),
    };

    this.onChange = this.onChange.bind(this);
  }

  onChange(prop, value) {
    const resistor = this.state.resistor;
    resistor.set(prop, value);
    this.setState({ resistor });
  }

  render() {
    return (
      <div className="resistor-calc">
        <h2>Resistor Calculator</h2>
        <h4>Calculates values from color codes</h4>
        <this.props.form model={this.state.resistor} onChange={this.onChange} />
        <ResistorValue
          className="resistor-value" formatOptions={{ short: true }}
          model={this.state.resistor}
        />
        <ResistorSvg model={this.state.resistor} />
      </div>
    );
  }
}

ResistorCalculator.propTypes = {
  form: PropTypes.func,
};

ResistorCalculator.defaultProps = {
  form: ResistorForm,
};

export default ResistorCalculator;
