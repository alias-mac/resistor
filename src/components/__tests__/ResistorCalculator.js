/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import ResistorCalculator from '../ResistorCalculator';
import ResistorForm from '../ResistorForm';

describe('ResistorCalculator', function () {
  it('should render with default state', function () {
    const component = renderer.create(
      <ResistorCalculator />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should update to 3-band resistor', function () {

    function MockedForm(props) {
      const { onChange } = props;

      return (
        <div name="mocked-form" onChange={onChange}>
        <ResistorForm {...props} />;
        </div>
      );
    }

    const component = renderer.create(
      <ResistorCalculator form={MockedForm} />
    );

    const form = component.toJSON().children.find(c => c.props.name === 'mocked-form');
    form.props.onChange('bands', '3');

    expect(component.toJSON()).toMatchSnapshot();
  });
});
