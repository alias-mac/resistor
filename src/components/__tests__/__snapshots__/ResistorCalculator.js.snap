exports[`ResistorCalculator should render with default state 1`] = `
<div
  className="resistor-calc">
  <h2>
    Resistor Calculator
  </h2>
  <h4>
    Calculates values from color codes
  </h4>
  <form
    className="resistor-form"
    onChange={[Function]}>
    <div
      className="row">
      <div
        className="form-group col-sm-6 offset-sm-3 col-md-2 offset-md-5">
        <label>
          Strips
        </label>
        <select
          className="form-control"
          name="bands"
          onChange={[Function]}
          value={4}>
          <option
            value={3}>
            3
            -strip
          </option>
          <option
            value={4}>
            4
            -strip
          </option>
          <option
            value={5}>
            5
            -strip
          </option>
          <option
            value={6}>
            6
            -strip
          </option>
        </select>
      </div>
    </div>
    <div
      className="row">
      <div
        className="col-xs-6 col-sm-2 offset-sm-2">
        <div
          className="form-group">
          <label>
            1st Digit
          </label>
          <select
            className="form-control brown"
            name="digit1"
            onChange={[Function]}
            value={1}>
            <option
              className="brown"
              label="brown"
              value={1}>
              brown
            </option>
            <option
              className="red"
              label="red"
              value={2}>
              red
            </option>
            <option
              className="orange"
              label="orange"
              value={3}>
              orange
            </option>
            <option
              className="yellow"
              label="yellow"
              value={4}>
              yellow
            </option>
            <option
              className="green"
              label="green"
              value={5}>
              green
            </option>
            <option
              className="blue"
              label="blue"
              value={6}>
              blue
            </option>
            <option
              className="violet"
              label="violet"
              value={7}>
              violet
            </option>
            <option
              className="gray"
              label="gray"
              value={8}>
              gray
            </option>
            <option
              className="white"
              label="white"
              value={9}>
              white
            </option>
          </select>
        </div>
      </div>
      <div
        className="col-xs-6 col-sm-2">
        <div
          className="form-group">
          <label>
            2nd Digit
          </label>
          <select
            className="form-control black"
            name="digit2"
            onChange={[Function]}
            value={0}>
            <option
              className="black"
              label="black"
              value={0}>
              black
            </option>
            <option
              className="brown"
              label="brown"
              value={1}>
              brown
            </option>
            <option
              className="red"
              label="red"
              value={2}>
              red
            </option>
            <option
              className="orange"
              label="orange"
              value={3}>
              orange
            </option>
            <option
              className="yellow"
              label="yellow"
              value={4}>
              yellow
            </option>
            <option
              className="green"
              label="green"
              value={5}>
              green
            </option>
            <option
              className="blue"
              label="blue"
              value={6}>
              blue
            </option>
            <option
              className="violet"
              label="violet"
              value={7}>
              violet
            </option>
            <option
              className="gray"
              label="gray"
              value={8}>
              gray
            </option>
            <option
              className="white"
              label="white"
              value={9}>
              white
            </option>
          </select>
        </div>
      </div>
      <div
        className="col-xs-6 col-sm-2">
        <div
          className="form-group">
          <label>
            Multiplier
          </label>
          <select
            className="form-control black"
            name="multiplier"
            onChange={[Function]}
            value={0}>
            <option
              className="black"
              label="black"
              value={0}>
              black
            </option>
            <option
              className="brown"
              label="brown"
              value={1}>
              brown
            </option>
            <option
              className="red"
              label="red"
              value={2}>
              red
            </option>
            <option
              className="orange"
              label="orange"
              value={3}>
              orange
            </option>
            <option
              className="yellow"
              label="yellow"
              value={4}>
              yellow
            </option>
            <option
              className="green"
              label="green"
              value={5}>
              green
            </option>
            <option
              className="blue"
              label="blue"
              value={6}>
              blue
            </option>
            <option
              className="violet"
              label="violet"
              value={7}>
              violet
            </option>
            <option
              className="gray"
              label="gray"
              value={8}>
              gray
            </option>
            <option
              className="white"
              label="white"
              value={9}>
              white
            </option>
            <option
              className="gold"
              label="gold"
              value={-1}>
              gold
            </option>
            <option
              className="silver"
              label="silver"
              value={-2}>
              silver
            </option>
          </select>
        </div>
      </div>
      <div
        className="col-xs-6 col-sm-2">
        <div
          className="form-group">
          <label>
            Tolerance
          </label>
          <select
            className="form-control silver"
            name="tolerance"
            onChange={[Function]}
            value={10}>
            <option
              className="brown"
              label="brown"
              value={1}>
              brown
            </option>
            <option
              className="red"
              label="red"
              value={2}>
              red
            </option>
            <option
              className="green"
              label="green"
              value={0.5}>
              green
            </option>
            <option
              className="blue"
              label="blue"
              value={0.25}>
              blue
            </option>
            <option
              className="violet"
              label="violet"
              value={0.1}>
              violet
            </option>
            <option
              className="gold"
              label="gold"
              value={5}>
              gold
            </option>
            <option
              className="silver"
              label="silver"
              value={10}>
              silver
            </option>
          </select>
        </div>
      </div>
    </div>
  </form>
  <p
    className="resistor-value">
    Resistance: 
    10 Ω ± 10%
  </p>
  <svg
    preserveAspectRatio="xMinYMin meet"
    viewBox="0 0 400 100"
    width="100%">
    <defs>
      <rect
        fill="#d6d6d6"
        height="100"
        id="edge"
        rx="10"
        ry="10"
        width="80" />
      <rect
        height="100%"
        id="stripe-wide"
        width="20" />
      <rect
        height={80}
        id="stripe"
        width="20"
        y="10" />
    </defs>
    <rect
      fill="#e8e8e8"
      height="24"
      width="100%"
      x="0"
      y={38} />
    <rect
      fill="#d6d6d6"
      height={80}
      width={280}
      x="60"
      y="10" />
    <use
      x="40"
      xlinkHref="#edge"
      y="0" />
    <use
      x={280}
      xlinkHref="#edge"
      y="0" />
    <use
      fill="brown"
      id="stripe-1"
      x="90"
      xlinkHref="#stripe-wide" />
    <use
      fill="black"
      id="stripe-2"
      x="130"
      xlinkHref="#stripe" />
    <use
      fill="black"
      id="multiplier"
      x="190"
      xlinkHref="#stripe" />
    <use
      fill="silver"
      id="tolerance"
      x="290"
      xlinkHref="#stripe-wide" />
  </svg>
</div>
`;
